src/main/java/InfrastructureManager/MatchMaking/ScoreBasedMM.java

```diff
@@ -0,0 +1,228 @@
+     */
+    @Override
+    public EdgeNode match(EdgeClient thisClient, List<EdgeNode> nodeListInput, EdgeClientHistory thisClientHistory) {
+        logger.info("""
+                
+                ----------------------------MATCH MAKING - SCORE BASED--------------------
+                R_weight, N_weight, D_weight, H_weight, QOS_threshold
+                {}      , {}      , {}      , {}      , {}    """
+                , RESOURCE_WEIGHT, NETWORK_WEIGHT, PING_WEIGHT, HISTORY_WEIGHT, QOS_THRESHOLD);
+
+        //Initiating variable
+        List<EdgeNode> nodeList = new ArrayList<>(nodeListInput); //In case of multi threading
+        EdgeNode bestNode = new EdgeNode();
+        EdgeNode rejectNode = new EdgeNode("rejectNode", "000.000.000.000", false, Long.MAX_VALUE, Long.MAX_VALUE, Long.MAX_VALUE, Long.MAX_VALUE, Long.MAX_VALUE); //return this if score not good
+        long bestScore = 0;
+
+        this.clientHistory = thisClientHistory;
+
+        //initiate temp/comparing variable
+
+        int numberOfUnqualified = 0; //to count the number of node we have been ruled out during iteration
+        int totalNumberOfNode = nodeList.size();
+
+        //Get client requirement (required resource, network)
+        String thisClientID = thisClient.getId();
+        long reqResource = thisClient.getReqResource();
+        long reqNetwork = thisClient.getReqNetwork();
+
+        //Get node's stats (ping, res, network, and history) and eliminate bad one
+        for (EdgeNode thisNode : nodeList) {
+            //List of comparing variables
+            String thisNodeID = thisNode.getId();
+            long nodeResource = thisNode.getResource();
+            long nodeNetwork = thisNode.getNetwork();
+            boolean nodeIsConnected = thisNode.isConnected();
+            long pingNumber = getPing(thisClient, thisNode);
+            long nodeHistoryWithClient = clientHistory.getHistoryScore(thisClientID, thisNodeID);
+            long lastConnectedTime = clientHistory.getConnectedTime(thisClientID, thisNodeID);
+
+            //Get elapsed time to reduce history score
+            long currentTime = System.currentTimeMillis();
+            long elapsedTime = currentTime - lastConnectedTime;
+
+            //Deducting score: -1 score for every second (1000ms)
+            long deductingScore = deductScorePerSecond * elapsedTime / 1000;
+            logger.info("time since last connected = {} ms, {} score deduct per second", elapsedTime, deductScorePerSecond);
+
+            long nodeHistoryScore = nodeHistoryWithClient - deductingScore;
+
+            if (nodeHistoryScore < 0) {
+                nodeHistoryScore = 0;
+            }
+            logger.info("node history score = {} = {} - {}", nodeHistoryScore, nodeHistoryWithClient, deductingScore);
+            //Update the score in the history info pack with this new score
+            clientHistory.setHistoryScoreForClient(thisClientID, thisNodeID, nodeHistoryScore);
```

Comment:
> Here you set the history score after deduction, but the connected time is still same
The only time that connectedTime is accessed is when we want to calculate how much time have passed in order to reduce the history score.  
So when it connected doesn't matter. We want the time it disconnected from the node only,

> That is going to happen if we have two assignments one after the other?
This is also what I fear, race condition may happen if we are not fast enough. Maybe synchronized would help in this case?

Reply:
It is not about race condition, only one thread is going to access one instance of match-making at a time.

The problem here can be reproduced with the following scenario:
- client disconnect with failure.
- assign client fails (no suitable node).
- if you call assign client again before history is 0, history value will be wrong.

I was able to reproduce it (after correcting another bug that I mentioned elsewhere).

The solution is either to set both history and time or neither.

Reply:
Done in 2c63281

The shared variable/object clientHistory still bug me somehow.
What if we have 10 instance, doing match making in the same times, then updating scores and history will result in race condition. 

Thinking about use semaphore or synchronize to avoid that situation. WDYT? @alyhasansakr